File: Sources/MicrosoftOrganizationChartReader/Records/BorderRecord.swift
//
//  BorderRecord.swift
//  
//
//  Created by Hugh Bellamy on 23/01/2021.
//

import DataStream

public struct BorderRecord {
    public let size: UInt16
    public let style: Style
    public let unknown1: UInt16
    public let unknown2: UInt16
    public let lineStyle: LineStyle
    public let borderColor: (red: UInt8, blue: UInt8, green: UInt8, unused: UInt8)
    public let unknown3: UInt16
    
    public init(dataStream: inout DataStream) throws {
        self.size = try dataStream.read(endianess: .littleEndian)
        self.style = try Style(dataStream: &dataStream)
        self.unknown1 = try dataStream.read(endianess: .littleEndian)
        self.unknown2 = try dataStream.read(endianess: .littleEndian)
        self.lineStyle = try LineStyle(dataStream: &dataStream)

        self.borderColor = (
            red: try dataStream.read(endianess: .littleEndian),
            blue: try dataStream.read(endianess: .littleEndian),
            green: try dataStream.read(endianess: .littleEndian),
            unused: try dataStream.read(endianess: .littleEndian)
        )
        
        self.unknown3 = try dataStream.read(endianess: .littleEndian)
    }
    
    public enum Style: UInt16, DataStreamCreatable {
        case none = 0x00
        case single1 = 0x01
        case hairline = 0x0A
        case single2 = 0x0C
        case single3 = 0x0D
        case single4 = 0x0E
        case double1 = 0x1F
        case double2 = 0x16
        case double3 = 0x29
        case double4 = 0x17
        case double5 = 0x33
        case double6 = 0x18
    }
}

File: Explorer/MicrosoftOrganizationChartReader/Records/BorderRecord.cs
using System.ComponentModel;
using System.Drawing;
using System.IO;

namespace OrganizationChartReader
{
    public class BorderRecord
    {
        public ushort Size { get; set; }

        public BorderStyle Style { get; set; }

        [TypeConverter(typeof(HexTypeConverter))]
        public ushort Unknown1 { get; set; }

        [TypeConverter(typeof(HexTypeConverter))]
        public ushort Unknown2 { get; set; }

        public LineStyle LineStyle { get; set; }

        public Color Color { get; set; }

        [TypeConverter(typeof(HexTypeConverter))]
        public ushort Unknown3 { get; set; }

        public BorderRecord(BinaryReader reader)
        {
            Size = reader.ReadUInt16();
            Style = (BorderStyle)reader.ReadUInt16();
            Unknown1 = reader.ReadUInt16();
            Unknown2 = reader.ReadUInt16();
            LineStyle = (LineStyle)reader.ReadUInt16();
            Color = reader.ReadOrganizationChartColor();
            Unknown3 = reader.ReadUInt16();
        }

        public enum BorderStyle : ushort
        {
            None = 0x00,
            Single1 = 0x01,
            Hairline = 0x0A,
            Single2 = 0x0C,
            Single3 = 0x0D,
            Single4 = 0x0E,
            Double1 = 0x1F,
            Double2 = 0x16,
            Double3 = 0x29,
            Double4 = 0x17,
            Double5 = 0x33,
            Double6 = 0x18
        }
    }
}
